drv/i2c-devices/src/pca9956b.rs

```diff
@@ -0,0 +1,332 @@
+        Error::I2cError(err)
+    }
+}
+
+impl From<Error> for ResponseCode {
+    fn from(err: Error) -> Self {
+        match err {
+            Error::I2cError(code) => code,
+            _ => panic!(),
+        }
+    }
+}
+
+impl Pca9956B {
+    pub fn new(device: &I2cDevice) -> Self {
+        Self { device: *device }
+    }
+
+    /// Read a single Register
+    fn read_reg(&self, reg: Register) -> Result<u8, Error> {
+        self.device
+            .read_reg::<u8, u8>(reg as u8)
+            .map_err(|code| Error::I2cError(code))
+    }
+
+    /// Read a number of Registers into `buf`
+    /// Note that CTRL_AUTO_INCR_MASK is set here to enable the device's
+    /// auto-increment feature, which has varying behaviors and limitations that
+    /// must be accounted for.
+    #[allow(dead_code)]
+    fn read_buffer(&self, reg: Register, buf: &mut [u8]) -> Result<(), Error> {
+        self.device
+            .read_reg_into((reg as u8) | CTRL_AUTO_INCR_MASK, buf)
+            .map_err(|code| Error::I2cError(code))?;
+
+        Ok(())
+    }
+
+    /// Write a single Register
+    fn write_reg(&self, reg: Register, val: u8) -> Result<(), Error> {
+        let buffer = [reg as u8, val];
+        self.device
+            .write(&buffer)
+            .map_err(|code| Error::I2cError(code))
+    }
+
+    /// Write a number of Registers into `buf`
+    /// Note that CTRL_AUTO_INCR_MASK is set here to enable the device's
+    /// auto-increment feature, which has varying behaviors and limitations that
+    /// must be accounted for.
+    fn write_buffer(&self, reg: Register, buf: &[u8]) -> Result<(), Error> {
+        let mut data: [u8; 25] = [0; 25];
+        data[0] = (reg as u8) | CTRL_AUTO_INCR_MASK;
+        data[1..=buf.len()].copy_from_slice(buf);
+
+        self.device
+            .write(&data[..=buf.len()])
+            .map_err(|code| Error::I2cError(code))
+    }
+
+    /// Sets the device's IREFALL register to `val`
+    /// IREFALL gets copied into all IREFn registers by the device.
+    pub fn set_iref_all(&self, val: u8) -> Result<(), Error> {
+        self.write_reg(Register::IREFALL, val)
+    }
+
+    /// Sets the device's PWMALL register to `val`
+    /// PWMALL gets copied into all PWMn registers by the device.
+    pub fn set_pwm_all(&self, val: u8) -> Result<(), Error> {
+        self.write_reg(Register::PWMALL, val)
+    }
+
+    /// Sets the PWMx register to `val` for a given `led`
+    pub fn set_a_led_pwm(&self, led: u8, val: u8) -> Result<(), Error> {
+        if led >= NUM_LEDS as u8 {
+            return Err(Error::InvalidLED(led));
+        }
+        let reg = FromPrimitive::from_u8((Register::PWM0 as u8) + led).unwrap();
+        self.write_reg(reg, val)
+    }
+
+    /// Sets the PWMx register a number of LEDs, up to NUM_LEDS, beginning with
+    /// PWM0.
+    pub fn set_all_led_pwm(&self, vals: &[u8]) -> Result<(), Error> {
+        if vals.len() > NUM_LEDS {
+            return Err(Error::InvalidLED(
```

Comment:
What about returning `Error::InvalidLED(NUM_LEDS)` (i.e. reporting the _first_ bad LED, instead of trying to report the _last_ bad LED and maybe falling back to 0xFF)?